### PR TITLE
Improve unicode support

### PR DESCRIPTION
**Pull request by bakkot**

### CODE REVIEW

File: scripts/make-identifier-regexen.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+const ID_Start = require('unicode-9.0.0/Binary_Property/ID_Start/code-points.js');
+const ID_Continue = require('unicode-9.0.0/Binary_Property/ID_Continue/code-points.js');
+const Other_ID_Start = require('unicode-9.0.0/Binary_Property/Other_ID_Start/code-points.js');
+
+
+function asFourHexDigits(cp) {
+  return cp.toString(16).padStart(4, 0);
```
Comment (michaelficarra) [inline scripts/make-identifier-regexen.js]: Implicit conversion of `0` to string. Gross.

Comment (bakkot) [reply scripts/make-identifier-regexen.js]: :| Had this right at some point...

Fixed.

File: scripts/make-identifier-regexen.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    } else {
+      alternatives.push(escape(cp));
+    }
+  }
+  return alternatives.join('|');
+}
+
+
+console.log('IdentifierStart:\n' + toRegex(ID_Start));
+console.log('IdentifierContinue:\n' + toRegex(Other_ID_Start.concat(ID_Continue, [0x200c, 0x200d])));
```
Comment (michaelficarra) [inline scripts/make-identifier-regexen.js]: Shouldn't this be `Other_ID_Continue`?

Comment (bakkot) [reply scripts/make-identifier-regexen.js]: [No.](https://github.com/estools/esutils/blob/master/tools/generate-identifier-regex.js#L56)

File: scripts/make-identifier-regexen.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    if (cp < 128) continue; // exclude ascii
+    if (array[i + 1] === cp + 1 && array[i + 2] === cp + 2) {
+      // i.e. we have a range
+      i += 2;
+      let offset = 2;
+      while (i < array.length && array[i + 1] === cp + offset + 1) {
+        ++i;
+        ++offset;
+      }
+      alternatives.push('[' + escape(cp) + '-' + escape(array[i]) + ']');
```
Comment (bakkot) [inline scripts/make-identifier-regexen.js]: It occurs to me that we could just have one giant character class with a bunch of ranges, rather than a bunch of character classes, and thereby have a much shorter regex. Should probably change that before this lands.

File: scripts/make-identifier-regexen.js (hunk 2 above)
Comment (michaelficarra) [reply scripts/make-identifier-regexen.js]: I'm not really sure I understand, but I'll go with it.

File: scripts/make-identifier-regexen.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  return '\\u' + asFourHexDigits(cu1) + '\\u' + asFourHexDigits(cu2);
+}
+
+function toRegex(array) {
+  const sorted = [...new Set(array)].sort((a, b) => a - b);
+  const classes = [];
+  for (let i = 0; i < sorted.length; ++i) {
+    const cp = sorted[i];
+    if (cp < 128) continue; // exclude ascii
+    if (sorted[i + 1] === cp + 1 && sorted[i + 2] === cp + 2) {
```
Comment (michaelficarra) [inline scripts/make-identifier-regexen.js]: This is a possible out-of-bounds access.

**Review by michaelficarra — APPROVED** [review]

**Review by michaelficarra — APPROVED** [review]

File: src/main/java/com/shapesecurity/shift/es2016/codegen/TokenStream.java [hunk 5]
```diff
@@ -67,17 +67,22 @@ public void put(@Nonnull String tokenStr) {
...
-        char rightChar = tokenStr.charAt(0);
-        char lastChar = this.lastChar;
-        this.lastChar = tokenStr.charAt(tokenStr.length() - 1);
-        if ((lastChar == '+' || lastChar == '-') && lastChar == rightChar ||
-                Utils.isIdentifierPart(lastChar) && Utils.isIdentifierPart(rightChar) ||
-                lastChar == '/' && (rightChar == 'i' || rightChar == '/')) {
+        int rightCodePoint = tokenStr.codePointAt(0);
+        int lastCodePoint = this.lastCodePoint;
+        char lastChar = tokenStr.charAt(tokenStr.length() - 1);
+        if (lastChar >= 0xDC00 && lastChar <= 0xDFFF) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2016/codegen/TokenStream.java:78]: We should probably make a helper in Utils like `isTrailSurrogate` (and `isLeadSurrogate`) and use it here and wherever else we perform this test -- probably at least the parser.

File: src/main/java/com/shapesecurity/shift/es2016/codegen/TokenStream.java [hunk 6]
```diff
@@ -67,17 +67,22 @@ public void put(@Nonnull String tokenStr) {
...
+            this.lastCodePoint = tokenStr.codePointAt(tokenStr.length() - 2);
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2016/codegen/TokenStream.java:79]: Should we be checking that the length is > 1?

File: src/main/java/com/shapesecurity/shift/es2016/codegen/TokenStream.java [hunk 7]
```diff
@@ -67,17 +67,22 @@ public void put(@Nonnull String tokenStr) {
...
         this.lastNumber = null;
-        char rightChar = tokenStr.charAt(0);
-        char lastChar = this.lastChar;
-        this.lastChar = tokenStr.charAt(tokenStr.length() - 1);
-        if ((lastChar == '+' || lastChar == '-') && lastChar == rightChar ||
-                Utils.isIdentifierPart(lastChar) && Utils.isIdentifierPart(rightChar) ||
-                lastChar == '/' && (rightChar == 'i' || rightChar == '/')) {
+        int rightCodePoint = tokenStr.codePointAt(0);
+        int lastCodePoint = this.lastCodePoint;
+        char lastChar = tokenStr.charAt(tokenStr.length() - 1);
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2016/codegen/TokenStream.java:77]: Should we be checking that the length is > 0?

File: src/main/java/com/shapesecurity/shift/es2016/codegen/TokenStream.java [hunk 8]
```diff
@@ -67,17 +67,22 @@ public void put(@Nonnull String tokenStr) {
...
+        if (lastChar >= 0xDC00 && lastChar <= 0xDFFF) {
+            this.lastCodePoint = tokenStr.codePointAt(tokenStr.length() - 2);
+        } else {
+            this.lastCodePoint = lastChar;
+        }
+        if ((lastCodePoint == '+' || lastCodePoint == '-') && lastCodePoint == rightCodePoint ||
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2016/codegen/TokenStream.java:83]: `lastCodePoint` is an integer, let's do integer comparisons. You can list the code point info in a comment above.

```java
// U+002B PLUS SIGN
// U+002D HYPHEN-MINUS
if ((lastCodePoint == 0x2B || lastCodePoint == 0x2D) && lastCodePoint == rightCodePoint ||
```

alternatively, we could use constants

```java
static final PLUS_SIGN = 0x2B; // U+002B PLUS SIGN
static final HYPHEN_MINUS = 0x2D; // U+002D HYPHEN-MINUS
// ...
if ((lastCodePoint == PLUS_SIGN || lastCodePoint == HYPHEN_MINUS) && lastCodePoint == rightCodePoint ||
```

File: src/test/java/com/shapesecurity/shift/es2016/parser/miscellaneous/WhitespaceTest.java [hunk 9]
```diff
@@ -98,7 +98,7 @@ public void testWhitespace() throws JsError {
 
         testScript("throw /* \u202a */ e", new ThrowStatement(new IdentifierExpression("e")));
 
-        testScript("new\u0020\u0009\u000B\u000C\u00A0\u1680\u180E\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u202F\u205F\u3000\uFEFFa",
+        testScript("new\u0020\u0009\u000B\u000C\u00A0\u1680\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u202F\u205F\u3000\uFEFFa",
```
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/es2016/parser/miscellaneous/WhitespaceTest.java:101]: This is a separate issue, updating Unicode data to recognise fewer whitespace characters. Please split it out into its own PR.
